### PR TITLE
Migrate backup-restore test from pulse to terraform

### PR DESCRIPTION
**Pull request by chrishajas**
Signed-off-by: Karen Huddleston <khuddleston@pivotal.io>
Signed-off-by: Chris Hajas <chajas@pivotal.io>

### CODE REVIEW

File: concourse/pipelines/pipeline.yml [hunk 1]
```diff
@@ -645,6 +687,53 @@ jobs:
...
+      - name: bin_gpdb
+      - name: ccp_src
+      - name: gpdb_src
+      run:
+        path: bash
+        args:
+        - -c
+        - |
+          set -ex
+          source ccp_src/aws/generate_env_files.sh
```
Comment (khuddlefish) [inline concourse/pipelines/pipeline.yml]: Maybe the setup functions before we run behave should be in their own task.

File: concourse/pipelines/pipeline.yml [hunk 2]
```diff
@@ -2,7 +2,49 @@
 ## resources
 ## ======================================================================
 
+resource_types:
```
Comment (larham) [inline concourse/pipelines/pipeline.yml]: consider putting resource_types under a separate header, such as 
```
## resource types
## =====================
```

File: concourse/pipelines/pipeline.yml [hunk 3]
```diff
@@ -2,7 +2,49 @@
...
+    repository: ljfranklin/terraform-resource
+
 resources:
+- name: ccp_src
+  type: git
+  source:
+    branch: {{ccp-git-branch}}
+    private_key: {{ccp-git-key}}
+    uri: {{ccp-git-remote}}
+    tag_filter: {{ccp-tag-filter}}
```
Comment (kmacoskey) [inline concourse/pipelines/pipeline.yml]: We talked this morning and decided to have everyone using latest until we have an actual breaking change that requires a release. Therefore, lets go ahead and remove the `tag_filter` and have `ccp_src` default to latest.

File: concourse/pipelines/pipeline.yml [hunk 4]
```diff
@@ -645,6 +687,53 @@ jobs:
...
+      trigger: true
+    - get: ccp_src
+    - get: centos-gpdb-dev-6
+  - put: terraform
+    params:
+      action: create
+      delete_on_failure: true
+      generate_random_name: true
+      <<: *ccp_options
+  - task: run_tests
```
Comment (kmacoskey) [inline concourse/pipelines/pipeline.yml]: Is there a particular reason you want this task to be defined inline? I like the visibility, but all other tasks in this pipeline are defined in separate task yml. Do we want visibility or consistency? Moving this task into a separate task yml would require repeating the `debug_sleep` anchor block in the task yml, but I'm ok with that.

File: concourse/pipelines/pipeline.yml (hunk 1 above)
Comment (kmacoskey) [inline concourse/pipelines/pipeline.yml]: Why are you sourcing this script? it should just be executed.

File: concourse/pipelines/pipeline.yml [hunk 5]
```diff
@@ -1358,3 +1429,32 @@ jobs:
         MAKE_TEST_COMMAND: optimizer_functional_part2
         BLDWRAP_POSTGRES_CONF_ADDONS: fsync=off optimizer_print_missing_stats=off
         TEST_OS: centos
+
+ccp_options_anchor: &ccp_options
```
Comment (larham) [inline concourse/pipelines/pipeline.yml]: consider putting these anchors under a separate header like "reusable anchors"

Also, I found it slightly odd to have the definition follow the usage. I guess it is nice to have details like this pushed down to the bottom, though.

File: concourse/scripts/run_behave_test.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+BEHAVE_FLAGS=$@
+
+run_behave_tests() {
+    local gpdb_host=$1
+
+    yum install -y openssh-clients
+    ssh -t $gpdb_host "bash -c \"\
```
Comment (larham) [inline concourse/scripts/run_behave_test.sh]: consider creating a file with this sequence of commands,  and using scp to copy that file onto the remote host, then running it via ssh.  The advantage of this extra work would be that during debugging, the command for the test will be evident and reusable.

File: concourse/pipelines/pipeline.yml [hunk 7]
```diff
@@ -1358,3 +1429,32 @@ jobs:
...
+  terraform_source: ccp_src/aws/
+  vars:
+    aws_instance-node-instance_type: t2.medium
```
Comment (kmacoskey) [inline concourse/pipelines/pipeline.yml]: Lets bump this up to the expected (same as pulse) `r3.xlarge` instance type.

File: concourse/pipelines/pipeline.yml [hunk 8]
```diff
@@ -1358,3 +1429,32 @@ jobs:
...
+    config:
+      platform: linux
+      image_resource:
+        type: docker-image
+        source:
+          repository: alpine
+          tag: latest
+      run:
+        path: 'sh'
+        args: ['-c', 'sleep 21600'] # 6 hours
```
Comment (kmacoskey) [inline concourse/pipelines/pipeline.yml]: The sleep in alpine is `BusyBox v1.26.2`, therefore you could do this as `sleep 6h`.

File: concourse/scripts/backup_utils.sh [hunk 9]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+copy_backup_files() {
+    for node in mdw sdw1 ; do
```
Comment (doty-pivotal) [inline concourse/scripts/backup_utils.sh]: So for now there are only two nodes, but if you have access to the master, might it be possible to query the DB for the number of hosts, and the host names that they use - thus making this script less dependent on the convention followed by the provisioning system? Something like this:

```
# on the master node
while read HOSTNAME; do
         ...
done <$(psql -At -c "select distinct hostname from gp_segment_configuration")
```

Its worth noting that in order for many of the install steps to go right, the master and the nodes do need to have the environment (ssh config, etc ...) set up correctly, and so relying on that may lessen the scripts dependance on the method it was called (from the concourse container for example).

Comment (doty-pivotal) [inline concourse/scripts/backup_utils.sh]: I find this site a nice place to look when I am trying to integrate bash with the DB: https://www.manniwood.com/postgresql_and_bash_stuff/index.html

File: concourse/pipelines/pipeline.yml (hunk 7 above)
Comment (divyabhargov) [inline concourse/pipelines/pipeline.yml]: I would not agree with @kmacoskey to necessarily make the instance type same as pulse. If the team thinks t2.medium is good enough to get the job done, its cool for it to stay so.

File: concourse/pipelines/pipeline.yml [hunk 10]
```diff
@@ -1358,3 +1429,32 @@ jobs:
...
+ccp_destroy_anchor: &ccp_destroy
+  put: terraform
+  params:
+    action: destroy
+    env_name_file: terraform/name
+    <<: *ccp_options
+  get_params:
+    action: destroy
+
+debug_sleep_anchor: &debug_sleep
```
Comment (jingyimei) [inline concourse/pipelines/pipeline.yml]: similar thinking here. Later on we may provide resource pooling tools to manage clusters for debugging and other scenarios, whose logic is hidden behind master pipeline. In that case, we don't need this debug and sleep step.

File: concourse/pipelines/pipeline.yml (hunk 2 above)
Comment (jingyimei) [inline concourse/pipelines/pipeline.yml]: It is good to keep them in consistent!

**Comment by kmacoskey** [comment]
We added a commit for a slight modification to the pipeline that will let you specify in your test what instance you want to use without requiring a test specific anchor. The `ccp_options` anchor has been completely removed.

FYI this is possible because the instance size passed to the terraform resource doesn't have to match in the `create` and `destroy`, the field just needs to exist. So the only anchor needed now is just `ccp_destroy` which will be identical for every team. 

We have not updated documentation or examples in CCP for this new pattern yet.

**Comment by chrishajas** [comment]
Thanks for the reviews! We've made a fair amount of changes to address the comments. We separated the behave task into a separate yaml file and call the script to run behave on the GPDB master instance instead of the concourse container.

**Comment by chrishajas** [comment]
Pushed! Thanks for the reviews!

**Pull request closed without merging**